### PR TITLE
#3528 Cant spread realm objects

### LINKED ISSUE #3528: Can't spread realm objects
## Describe the bug

Realm objects can't be spread when using V8(?) in development. We have a few cases where this is used in the app which need to be updated.

### To reproduce

Internal code issue

### Expected behaviour

No realm spreading!

### Proposed Solution

Search for `...` and if it's a realm object.. change it 🤷 

### Version and device info

- App version: 7.0.0-rc3
- Tablet model:
- OS version:

### Additional context

N/A

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #3528 #3541 

## Change summary

- Can't spread realm objects, so I did a global search and searched for `...` and then checked if they were a a realm object and changed it. Was super fun

### Prescriptions
- [x] Can select an insurance policy when creating a prescription
- [x] Can select a payment type when creating a prescription
- [x] Selecting a prescription category when creating a prescription
- [x] Can select a patient type when creating a prescription
- [x] Can edit a comment when creating a prescription
- [x] Can select a prescriber when creating a prescription
- [ ] When selecting an INACTIVE insurance policy, there is no discount applied

### Supplier requisitions
- [ ] Can apply a reason to a supplier requisition line
- [x] Can edit the comment on a SR

### Customer invoice
- [x] Can edit the their ref
- [x] Can edit the comment on a CI

### Supplier invoice
- [ ] Can apply a VVM status to a supplier invoice line
- [x] Can edit the batch name of a line in a supplier invoice
- [x] Can edit the selling price in a supplier invoice
- [x] Can edit the their ref
- [x] Can edit the comment on a SI

### Stocktake
- [x] Can apply a VVM status to a stocktake batch
- [x] Can edit a supplier of a stocktake batch
- [x] Can apply a reason to a stocktake batch
- [x] Can edit the sell price of a stocktake batch
- [x] Can edit the name of a stocktake (from the stocktake edit page)
- [x] Can edit the comment on a stocktake

### Customer requisition
- [x] Can edit a customer requisitions created date (program CR)
- [x] Can edit the comment on a CR

### Related areas to think about

N/A

### CODE REVIEW

File: src/actions/PrescriptionActions.js [hunk 1]
```diff
@@ -39,12 +39,10 @@ const cancelPrescription = () => (dispatch, getState) => {
...
   const { prescription } = getState();
   const { transaction } = prescription;
+  const { id } = transaction;
 
   UIDatabase.write(() => {
-    UIDatabase.update('Transaction', {
-      ...transaction,
-      category: newValue,
-    });
+    UIDatabase.update('Transaction', { id, category: newValue });
```
Comment (joshxg) [inline src/actions/PrescriptionActions.js:45]: This is selecting a prescription category when creating a prescription

File: src/actions/PrescriptionActions.js [hunk 2]
```diff
@@ -53,12 +51,10 @@ const editTransactionCategory = newValue => (dispatch, getState) => {
...
   const { prescription } = getState();
   const { transaction } = prescription;
+  const { id } = transaction;
 
   UIDatabase.write(() => {
-    UIDatabase.update('Transaction', {
-      ...transaction,
-      user1: newValue,
-    });
+    UIDatabase.update('Transaction', { id, user1: newValue });
```
Comment (joshxg) [inline src/actions/PrescriptionActions.js:57]: selecting a patient type when creating a prescription

File: src/actions/PrescriptionActions.js [hunk 3]
```diff
@@ -67,14 +63,11 @@ const editPatientType = newValue => (dispatch, getState) => {
...
-  const { comment } = transaction;
+  const { id, comment } = transaction;
 
   if (newValue !== comment) {
     UIDatabase.write(() => {
-      UIDatabase.update('Transaction', {
-        ...transaction,
-        comment: newValue,
-      });
+      UIDatabase.update('Transaction', { id, comment: newValue });
```
Comment (joshxg) [inline src/actions/PrescriptionActions.js:70]: editing a comment when creating a prescription

File: src/actions/PrescriptionActions.js [hunk 4]
```diff
@@ -173,12 +166,10 @@ const editQuantity = (id, quantity) => (dispatch, getState) => {
...
   const { transaction } = prescription;
+  const { id } = transaction;
+
   UIDatabase.write(() => {
-    UIDatabase.update('Transaction', {
-      // Previously used spread, but Realm.Objects don't like being spread it turns out. Issue#3323
-      id: transaction.id,
-      prescriber,
-    });
+    UIDatabase.update('Transaction', { id, prescriber });
```
Comment (joshxg) [inline src/actions/PrescriptionActions.js:172]: selecting a prescriber when creating a prescription

File: src/database/DataTypes/RequisitionItem.js [hunk 5]
```diff
@@ -155,9 +155,7 @@ export class RequisitionItem extends Realm.Object {
       .join(' OR ')}) AND item = $0`;
 
     // Return the maximum price of all MasterListItems.
-    return UIDatabase.objects('MasterListItem')
-      .filtered(queryString, this.item)
-      .max('price');
+    return UIDatabase.objects('MasterListItem').filtered(queryString, this.item).max('price');
```
Comment (joshxg) [inline src/database/DataTypes/RequisitionItem.js:158]: huh

File: src/database/DataTypes/RequisitionItem.js [hunk 6]
```diff
@@ -207,10 +205,7 @@ export class RequisitionItem extends Realm.Object {
   applyReason(database, newOption) {
     if (this.hasVariance) {
       database.write(() => {
-        database.update('RequisitionItem', {
-          ...this,
-          option: newOption,
-        });
+        database.update('RequisitionItem', { id: this.id, option: newOption });
```
Comment (joshxg) [inline src/database/DataTypes/RequisitionItem.js:208]: Applying a reason to a supplier requisition line

File: src/database/DataTypes/StocktakeBatch.js [hunk 7]
```diff
@@ -264,7 +264,7 @@ export class StocktakeBatch extends Realm.Object {
 
     database.write(() => {
       database.update('StocktakeBatch', {
-        ...this,
+        id: this.id,
```
Comment (joshxg) [inline src/database/DataTypes/StocktakeBatch.js:267]: Applying a reason to a stocktake batch

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 8]
```diff
@@ -50,9 +50,10 @@ export const editSensorName = (newValue, rowKey, route) => (dispatch, getState)
   const { data, keyExtractor } = selectPageState(getState());
 
   const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
+  const { id } = objectToEdit;
 
   if (objectToEdit) {
-    UIDatabase.write(() => UIDatabase.update('Sensor', { ...objectToEdit, name: newValue }));
+    UIDatabase.write(() => UIDatabase.update('Sensor', { id, name: newValue }));
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:56]: uhh this should actually be removed.. hm

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 9]
```diff
@@ -61,11 +62,10 @@ export const editLocationDescription = (newValue, rowKey, route) => (dispatch, g
   const { data, keyExtractor } = selectPageState(getState());
 
   const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
+  const { id } = objectToEdit;
 
   if (objectToEdit) {
-    UIDatabase.write(() =>
-      UIDatabase.update('Location', { ...objectToEdit, description: newValue })
-    );
+    UIDatabase.write(() => UIDatabase.update('Location', { id, description: newValue }));
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:68]: uhh this should actually be removed

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 10]
```diff
@@ -74,9 +74,10 @@ export const editLocationCode = (newValue, rowKey, route) => (dispatch, getState
   const { data, keyExtractor } = selectPageState(getState());
 
   const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
+  const { id } = objectToEdit;
 
   if (objectToEdit) {
-    UIDatabase.write(() => UIDatabase.update('Location', { ...objectToEdit, code: newValue }));
+    UIDatabase.write(() => UIDatabase.update('Location', { id, code: newValue }));
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:80]: uhh this should actually be removed

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 11]
```diff
@@ -94,9 +95,10 @@ export const editBatchDoses = (newValue, rowKey, route) => (dispatch, getState)
 
 export const editBatchVvmStatus = (vvmStatus, objectType, route) => (dispatch, getState) => {
   const { modalValue, keyExtractor } = selectPageState(getState());
+  const { id = '' } = modalValue ?? {};
 
   UIDatabase.write(() =>
-    UIDatabase.update(objectType, { ...modalValue, vaccineVialMonitorStatus: vvmStatus })
+    UIDatabase.update(objectType, { id, vaccineVialMonitorStatus: vvmStatus })
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:101]: Applying a VVM status to either a stocktkae batch or supplier invoice line

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 12]
```diff
@@ -241,10 +243,12 @@ export const editNegativeAdjustments = (value, rowKey, route) => (dispatch, getS
 export const editBatchName = (value, rowKey, objectType, route) => (dispatch, getState) => {
   const { data, keyExtractor } = selectPageState(getState());
   const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
+  const { id = '' } = objectToEdit ?? {};
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:246]: Editing a batch name in an SI or stocktkae

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 13]
```diff
@@ -254,23 +258,21 @@ export const editSellPrice = (value, rowKey, route) => (dispatch, getState) => {
   const { data, keyExtractor } = selectPageState(getState());
   const objectToEdit = data.find(row => keyExtractor(row) === rowKey);
   const objectDataType = objectToEdit.stocktake ? 'StocktakeBatch' : 'TransactionBatch';
-  const { sellPrice } = objectToEdit;
+  const { sellPrice, id } = objectToEdit;
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:261]: Editing the sell price in si or stocktake

File: src/pages/dataTableUtilities/actions/cellActions.js [hunk 14]
```diff
@@ -254,23 +258,21 @@ export const editSellPrice = (value, rowKey, route) => (dispatch, getState) => {
...
     UIDatabase.write(() => {
-      UIDatabase.update(objectDataType, { ...objectToEdit, sellPrice: currencyValue });
+      UIDatabase.update(objectDataType, { id, sellPrice: currencyValue });
       dispatch(refreshRow(rowKey, route));
     });
   }
 };
 
 export const editBatchSupplier = (supplier, batch, route) => dispatch => {
+  const { id } = batch;
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/cellActions.js:273]: editing the supplier in a stocktake

File: src/pages/dataTableUtilities/actions/pageActions.js [hunk 15]
```diff
@@ -103,25 +103,21 @@ export const closeDatePicker = route => ({
...
 export const editCreatedDate = (value, route) => (dispatch, getState) => {
   const pageObject = selectPageObject(getState());
+  const { id } = pageObject;
 
   UIDatabase.write(() => {
-    UIDatabase.update('Requisition', {
-      ...pageObject,
-      createdDate: value,
-    });
+    UIDatabase.update('Requisition', { id, createdDate: value });
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/pageActions.js:109]: editing a customer requisition created date

File: src/pages/dataTableUtilities/actions/pageActions.js [hunk 16]
```diff
@@ -103,25 +103,21 @@ export const closeDatePicker = route => ({
...
 export const editPrescriber = (value, route) => (dispatch, getState) => {
   const pageObject = selectPageObject(getState());
+  const { id } = pageObject;
 
   UIDatabase.write(() => {
-    UIDatabase.update('Transaction', {
-      ...pageObject,
-      prescriber: value,
-    });
+    UIDatabase.update('Transaction', { id, prescriber: value });
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/pageActions.js:120]: selecting a prescriber in a prescriber

File: src/pages/dataTableUtilities/actions/pageActions.js [hunk 17]
```diff
@@ -136,14 +132,11 @@ export const editPrescriber = (value, route) => (dispatch, getState) => {
...
-  const { name } = pageObject;
+  const { id, name } = pageObject;
 
   if (name !== value) {
     UIDatabase.write(() => {
-      UIDatabase.update(pageObjectType, {
-        ...pageObject,
-        name: value,
-      });
+      UIDatabase.update(pageObjectType, { id, name: value });
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/pageActions.js:139]: editing a stocktake name

File: src/pages/dataTableUtilities/actions/pageActions.js [hunk 18]
```diff
@@ -159,11 +152,11 @@ export const editPageObjectName = (value, pageObjectType, route) => (dispatch, g
 export const editTheirRef = (value, pageObjectType, route) => (dispatch, getState) => {
   const pageObject = selectPageObject(getState());
 
-  const { theirRef } = pageObject;
+  const { theirRef, id } = pageObject;
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/pageActions.js:155]: editing their ref on a transaction

File: src/pages/dataTableUtilities/actions/pageActions.js [hunk 19]
```diff
@@ -179,11 +172,11 @@ export const editTheirRef = (value, pageObjectType, route) => (dispatch, getStat
 export const editComment = (value, pageObjectType, route) => (dispatch, getState) => {
   const pageObject = selectPageObject(getState());
 
-  const { comment } = pageObject;
+  const { comment, id } = pageObject;
 
   if (comment !== value) {
     UIDatabase.write(() => {
-      UIDatabase.update(pageObjectType, { ...pageObject, comment: value });
+      UIDatabase.update(pageObjectType, { id, comment: value });
```
Comment (joshxg) [inline src/pages/dataTableUtilities/actions/pageActions.js:179]: editing a transaction comment

File: src/selectors/patient.js [hunk 20]
```diff
@@ -34,12 +34,15 @@ export const selectCurrentPatient = ({ patient }) => {
...
 export const selectPatientInsurancePolicies = ({ patient }) => {
   const { currentPatient } = patient;
   const { policies } = currentPatient;
   return policies.map(policy => {
-    const { isActive, policyNumber } = policy;
-    return isActive ? policy : { ...policy, policyNumber: `${policyNumber} (inactive)` };
+    const { isActive, policyNumber, id, discountRate } = policy;
+    return isActive
+      ? policy
+      : { isActive, id, discountRate, policyNumber: `${policyNumber} (inactive)` };
```
Comment (joshxg) [inline src/selectors/patient.js:45]: This is when you're viewing an insurance policy, so selecting on and seeing the deets are correct

File: src/widgets/PrescriptionInfo.js [hunk 21]
```diff
@@ -82,7 +82,7 @@ const PrescriptionInfoComponent = ({
             )}
           </FlexView>
 
-          <FlexRow flex={1} justifyContent="">
+          <FlexRow flex={1}>
```
Comment (joshxg) [inline src/widgets/PrescriptionInfo.js:85]: fixing #3541

File: src/pages/dataTableUtilities/actions/cellActions.js (hunk 8 above)
Comment (joshxg) [reply src/pages/dataTableUtilities/actions/cellActions.js:56]: I think I will make another PR for the stuff to be removed.. cleaner/easier to review? Let me know what you want!

**Comment by mark-prins** [comment]
> Was super fun

...misinformation (attempt at a visual word pun)

File: src/widgets/PrescriptionInfo.js (hunk 21 above)
Comment (mark-prins) [reply src/widgets/PrescriptionInfo.js:85]: thanks - testing this PR was tricky without this fix :)

**Comment by mark-prins** [comment]
The inactive insurance policy still gives me a discount. I'll have a look.

**Review by mark-prins — APPROVED** [review]
code changes look great. who knew there were so many latent bugs?!

I can't get vvm to show on supplier invoices or reasons on supplier requisitions but I'm sure both are fine. The insurance discount for inactive policies looks to be a problem outside of this issue.

**Comment by joshxg** [comment]
Will make a new PR, gonna get this into an RC. Thanks!

**Pull request merged**